File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {

    //     int[] ans = new int[k];
    //    PriorityQueue<List<Integer>> pq = new PriorityQueue<>((list1, list2) -> Integer.compare(list2.get(1), list1.get(1)));
    //    Map<Integer, Integer> mp = new HashMap<>();

    //    for (int x: nums) mp.put(x,mp.getOrDefault(x,0)+1);

    //    for (Map.Entry<Integer, Integer> etr: mp.entrySet()){
    //        int num = etr.getKey();
    //        int freq = etr.getValue();

    //        pq.offer(Arrays.asList(num,freq));
    //    }


    //    for (int i = 0;i<k;i++){
    //        ans[i] = pq.poll().get(0);
    //    }

    //    return ans;


       int[] ans = new int[k];
       
       Map<Integer, Integer> mp = new HashMap<>();

       for (int x: nums) mp.put(x,mp.getOrDefault(x,0)+1);
       int maxcount = 0;
       for (Map.Entry<Integer,Integer> etr: mp.entrySet()){
           maxcount = Math.max(maxcount,etr.getValue());
       }
       List<List<Integer>> counter = new ArrayList<>();
       for (int i = 0;i<maxcount;i++){
           counter.add(new ArrayList<Integer>());
       }
       
       for (Map.Entry<Integer,Integer> etr: mp.entrySet()){
           int val = etr.getKey();
           int freq = etr.getValue();
           counter.get(freq-1).add(val);
           
       }
       
       int j = 0;
       for (int i = counter.size()-1;i >=0;i--){
           if (k == 0) break;

           List<Integer> cur = counter.get(i);
           for (int x: cur){
               if (k == 0) return ans;
               ans[j++] = x;
               k--;
           }
       }

       return ans;
       

       





    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        mp = {}
        sorted_list = []
        ans = []
        for num in nums:
            mp[num] = mp.get(num, 0) + 1

        
        # for key,v in mp.items():
        #     sorted_list.append((v,key))

        # sorted_list.sort(key = lambda x: x[0], reverse = True)

        # for i in range(k):
        #     ans.append(sorted_list[i][1])

        # return ans
        pq = []

        for num, counts in mp.items():
            
            heapq.heappush(pq, (counts, num))
            if len(pq) > k:
                heapq.heappop(pq)

        for i in range(k):
            ans.append(heapq.heappop(pq)[1])

        return ans
